File: cs50/pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    // Variable declarations
    int last, first, firstTwo, count, length, multiply, digit0, digit, digit1, luhnTotal, test;
    long cc, ccCopy, ccCopy1;
    int multSum = 0;
    int addSum = 0;
    // Variable initializations
    cc = get_long("Enter your credit card number: \n");
    length = log10(cc) + 1;
    count = log10(cc);
    first =  cc / pow(10, count);
    firstTwo = cc / pow(10, count - 1);
    ccCopy = cc;
    ccCopy1 = cc;

    // Luhn Stage #1
    for (int i = length - 1; i >= 0; i -= 2)
    {
        digit = (ccCopy / 10) % 10;
        digit0 = digit * 2;

        test = log10(digit0);

        if (test == 1)
        {
            digit = (digit0 % 10) + ((digit0 / 10) % 10);
            multSum += digit;
            ccCopy /= 10;
            ccCopy /= 10;
        }
        else
        {
            multSum += (digit * 2);
            ccCopy /= 10;
            ccCopy /= 10;
        }
    }
    // Luhn Stage #1a
    for (int j = length - 1; j >= 0; j -= 2)
    {
        digit1 = ccCopy1 % 10;
        addSum += digit1;
        ccCopy1 /= 10;
        ccCopy1 /= 10;
    }
    // Luhn Stage #2
    luhnTotal = addSum + multSum;
    // If CC passes Luhn's Algorithm -- then get the card type!
    if (luhnTotal % 10 == 0)
    {
        if (((firstTwo == 34) || (firstTwo == 37)) && (length == 15))
        {
            printf("AMEX\n");
        }
        else if ((first == 4) && ((length == 13) || (length == 16)))
        {
            printf("VISA\n");
        }
        else if (((firstTwo == 51) || (firstTwo == 52) || (firstTwo == 53) || (firstTwo == 54) || (firstTwo == 55)) && length == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: cs50/pset6/credit.py
# training wheels
from cs50 import get_string

# variable declarations
# stripping any whitespace at the beginning or end
cc = get_string("Enter your credit card number: ").strip()
# variables for luhns algorithm
multSum = addSum = luhnsTotal = 0
# variables for length and first / first two digits of card for card type detection
length, first, firstTwo = len(cc), int(cc[0]), int(cc[0] + cc[1])

# initial for loop for the multiplication part of luhns
for i in range((len(cc) - 2), -1, -2):
    # multiple every other digit by 2 starting with second to last
    digit = (int(cc[i]) * 2)
    # check if the result is more than one digit long
    isMultiDigit = len(str(abs(digit)))
    # if more than one digit long -> add the two digits together -> add that to our multSum
    if(isMultiDigit == 2):
        digit = int(str(digit)[0]) + int(str(digit)[1])
        multSum += digit
    # otherwise just add the result to the sum
    else:
        multSum += digit

# simpler for loop to add up the non multiplied numbers
for j in range(len(cc) - 1, -1, -2):
        addSum += int(cc[j])

# add both totals together
luhnsTotal = multSum + addSum

# check which card type it is based on length and first one or two digits
if (luhnsTotal % 10) == 0:
    if (firstTwo == 34 or firstTwo == 37) and length == 15:
        print("AMEX")
    elif first == 4 and (length == 13 or length == 16):
        print("VISA")
    elif (firstTwo == 51 or firstTwo == 52 or firstTwo == 53 or firstTwo == 54 or firstTwo == 55) and length == 16:
        print("MASTERCARD")
    # if no digit / length combo matches print invalid
    else:
        print("INVALID")
# if fails luhns test print invalid
else:
    print("INVALID")
